### PR TITLE
Yaml config files

### LINKED ISSUE #1190: YAML configuration files
Add support for .yml configuration files. Simple as such.

### PR DESCRIPTION
**Pull request by AffectedArc07**
:cl: Core
The new configuration version is 2.3.0.
Configuration files are now stored in YAML format. Existing JSON config files will still be loaded with legacy support.
/:cl:

Closes #1190 

Example of yaml config
```yml
Database:
  DatabaseType: MariaDB
  ResetAdminPassword: false
  ConnectionString: lolno
  DropDatabase: false
  DesignTime: false
  ServerVersion: 10.5.5
General:
  ConfigVersion:
    Major: 2
    Minor: 3
    Build: 0
    Revision: -1
    MajorRevision: -1
    MinorRevision: -1
  ApiPort: 5000
  GitHubAccessToken: 
  SetupWizardMode: Never
  ByondTopicTimeout: 5000
  RestartTimeout: 60000
  UseBasicWatchdog: false
  HostApiDocumentation: true
  MinimumPasswordLength: 5
  InstanceLimit: 10
  UserLimit: 100
  UserGroupLimit: 25
  ValidInstancePaths: 
FileLogging:
  Directory: 
  Disable: false
  LogLevel: Debug
  MicrosoftLogLevel: Warning
ControlPanel:
  Enable: false
  AllowAnyOrigin: true
  AllowedOrigins: 
Swarm: 
```

### CODE REVIEW

**Comment by AffectedArc07** [comment]
> will decrease coverage by 86.61%.
fucking what?

**Comment by AffectedArc07** [comment]
> will decrease coverage by 0.04%

thats better

File: src/Tgstation.Server.Host/ServerFactory.cs [hunk 1]
```diff
@@ -49,7 +49,11 @@ public async Task<IServer> CreateServer(string[] args, string updatePath, Cancel
 
 			var basePath = IOManager.ResolvePath();
 			IHostBuilder CreateDefaultBuilder() => Microsoft.Extensions.Hosting.Host.CreateDefaultBuilder(args)
-				.ConfigureAppConfiguration((context, configuration) => configuration.SetBasePath(basePath));
```
Comment (AffectedArc07) [inline src/Tgstation.Server.Host/ServerFactory.cs:52]: Unsure if removing `configuration.SetBasePath()` here is gonna break stuff. It didnt in my testing but who knows.

File: README.md [hunk 2]
```diff
@@ -78,27 +78,27 @@ Note that this container is meant to be long running. Updates are handled intern
...
 
 #### Manual Configuration
 
-Create an `appsettings.Production.json` file next to `appsettings.json`. This will override the default settings in appsettings.json with your production settings. There are a few keys meant to be changed by hosts. Modifying any config files while the server is running will trigger a safe restart (Keeps DreamDaemon instances running). Note these are all case-sensitive:
+Create an `appsettings.Production.yml` file next to `appsettings.json`. This will override the default settings in `appsettings.json` with your production settings. There are a few keys meant to be changed by hosts. Modifying any config files while the server is running will trigger a safe restart (Keeps DreamDaemon instances running). Note these are all case-sensitive:
 
 - `General:ConfigVersion`: Suppresses warnings about out of date config versions. You should change this after updating TGS to one with a new config version. The current version can be found on the releases page for your server version (This field did not exist before v4.4.0).
```
Comment (Cyberboss) [inline README.md:104]: Can you update the oauth config example below please?

File: src/Tgstation.Server.Host/ServerFactory.cs [hunk 3]
```diff
@@ -49,7 +49,11 @@ public async Task<IServer> CreateServer(string[] args, string updatePath, Cancel
...
+				.ConfigureAppConfiguration((context, builder) => {
+					builder.AddJsonFile($"appsettings.{context.HostingEnvironment.EnvironmentName}.json", optional: true, reloadOnChange: true)
+					.AddYamlFile($"appsettings.{context.HostingEnvironment.EnvironmentName}.yml", optional: true, reloadOnChange: true); //Allow both appsettings.json and appsettings.yml for backwards compat
```
Comment (Cyberboss) [inline src/Tgstation.Server.Host/ServerFactory.cs:54]: The .json file is already configured by [CreateDefaultBuilder()](https://github.com/dotnet/runtime/blob/6a5a78bec9a6e14b4aa52cd5ac558f6cf5c6a211/src/libraries/Microsoft.Extensions.Hosting/src/Host.cs#L76). We shouldn't need to redo it.

```suggestion
					builder.AddYamlFile($"appsettings.yml", optional: true, reloadOnChange: true)
					.AddYamlFile($"appsettings.{context.HostingEnvironment.EnvironmentName}.yml", optional: true, reloadOnChange: true); //Allow both appsettings.json and appsettings.yml for backwards compat
```

File: src/Tgstation.Server.Host/Setup/SetupWizard.cs [hunk 4]
```diff
@@ -841,8 +842,11 @@ async Task<Uri> ParseAddress(string question)
 				{ SwarmConfiguration.Section, swarmConfiguration },
 			};
 
-			var json = JsonConvert.SerializeObject(map, Formatting.Indented);
-			var configBytes = Encoding.UTF8.GetBytes(json);
```
Comment (Cyberboss) [inline src/Tgstation.Server.Host/Setup/SetupWizard.cs:845]: Can you check and see what is still `using Newtonsoft.Json;` in this file? It should be nothing at this point.

**Review by Cyberboss — CHANGES_REQUESTED** [review]
- Redo the base `appsettings.json` as yaml please

**Comment by AffectedArc07** [comment]
Still works with `appsettings.json` legacy files

**Review by Cyberboss — CHANGES_REQUESTED** [review]
I noticed in your sample config the `General:ConfigVersion` property got fully serialized as not-semver. Can you fix that please?

File: src/Tgstation.Server.Host/ServerFactory.cs [hunk 5]
```diff
@@ -49,7 +49,10 @@ public async Task<IServer> CreateServer(string[] args, string updatePath, Cancel
 
 			var basePath = IOManager.ResolvePath();
 			IHostBuilder CreateDefaultBuilder() => Microsoft.Extensions.Hosting.Host.CreateDefaultBuilder(args)
-				.ConfigureAppConfiguration((context, configuration) => configuration.SetBasePath(basePath));
+				.ConfigureAppConfiguration((context, builder) => {
+					builder.AddYamlFile($"appsettings.{context.HostingEnvironment.EnvironmentName}.yml", optional: true, reloadOnChange: true); //Allow both appsettings.json and appsettings.yml for backwards compat
```
Comment (Cyberboss) [inline src/Tgstation.Server.Host/ServerFactory.cs:53]:
```suggestion
					builder.SetBasePath(basePath);
					builder.AddYamlFile($"appsettings.{context.HostingEnvironment.EnvironmentName}.yml", optional: true, reloadOnChange: true); //Allow both appsettings.json and appsettings.yml for backwards compat
```

9001% sure this is necessary for properly locating the files in production

**Review by Cyberboss — CHANGES_REQUESTED** [review]

File: src/Tgstation.Server.Host/ServerFactory.cs (hunk 5 above)
Comment (AffectedArc07) [reply src/Tgstation.Server.Host/ServerFactory.cs:53]: It seemed to locate just fine (I started it with `dotnet Tgstation.Server.Host.dll`, not the debugger), but ill add the base yml

**Comment by Cyberboss** [comment]
Try building a production environment and testing it for that base config change
```sh
cd src/Tgstation.Server.Host.Console
dotnet publish -c Release -o ../../prodtest/
cd ../Tgstation.Server.Host
dotnet publish -c Release --no-build -o ../../prodtest/lib/Default
```

**Comment by Cyberboss** [comment]
Also whatever you did broke a unit test

**Comment by AffectedArc07** [comment]
Cannot build host console
```
C:\Users\AffectedArc07\Documents\GitHub\tgstation-server\src\Tgstation.Server.Host.Console>dotnet publish -c Release -o ../../prodtest/
Microsoft (R) Build Engine version 16.8.0+126527ff1 for .NET
Copyright (C) Microsoft Corporation. All rights reserved.

  Determining projects to restore...
  All projects are up-to-date for restore.
CSC : error CS8034: Unable to load Analyzer assembly C:\Users\AffectedArc07\.nuget\packages\microsoft.codeanalysis.netanalyzers\5.0.1\analyzers\dotnet\cs\Microsoft.CodeAnalysis.CSharp.NetAnalyzers.dll : Assembly with same name is already loaded [C:\Users\AffectedArc07\Documents\GitHub\tgstation-server\src\Tgstation.Server.Host.Watchdog\Tgstation.Server.Host.Watchdog.csproj]
CSC : error CS8034: Unable to load Analyzer assembly C:\Users\AffectedArc07\.nuget\packages\microsoft.codeanalysis.netanalyzers\5.0.1\analyzers\dotnet\cs\Microsoft.CodeAnalysis.NetAnalyzers.dll : Assembly with same name is already loaded [C:\Users\AffectedArc07\Documents\GitHub\tgstation-server\src\Tgstation.Server.Host.Watchdog\Tgstation.Server.Host.Watchdog.csproj]

```

**Comment by AffectedArc07** [comment]
Seems to work locally, but would like you to look over it in detail. I dont wanna break it all.

**Comment by AffectedArc07** [comment]
Found a lead
```cs
var result = await factory.CreateServer(new[] { "General:SetupWizardMode=Never" }, Path, default);
```
Whenever `SetupWizardMode=Never` is set, the tests OOM and break.

File: src/Tgstation.Server.Host/ServerFactory.cs [hunk 6]
```diff
@@ -49,7 +49,11 @@ public async Task<IServer> CreateServer(string[] args, string updatePath, Cancel
 
 			var basePath = IOManager.ResolvePath();
 			IHostBuilder CreateDefaultBuilder() => Microsoft.Extensions.Hosting.Host.CreateDefaultBuilder(args)
-				.ConfigureAppConfiguration((context, configuration) => configuration.SetBasePath(basePath));
+				.ConfigureAppConfiguration((context, builder) => {
+					builder.AddYamlFile("appsettings.yml", optional: true, reloadOnChange: true)
+					.AddYamlFile($"appsettings.{context.HostingEnvironment.EnvironmentName}.yml", optional: true, reloadOnChange: true);
```
Comment (Cyberboss) [inline src/Tgstation.Server.Host/ServerFactory.cs:54]: To fix it we'll have to reorganize the builder's sources after creating them

```suggestion
					builder.AddYamlFile("appsettings.yml", optional: true, reloadOnChange: true)
						.AddYamlFile($"appsettings.{context.HostingEnvironment.EnvironmentName}.yml", optional: true, reloadOnChange: true);

					// reorganize the builder so our yaml configs don't override the env/cmdline configs
					// values obtained via debugger
					var envConfig = builder.Sources[3];
					var cmdLineConfig = builder.Sources[4];
					var baseYmlConfig = builder.Sources[5];
					var environmentYmlConfig = builder.Sources[6];
					builder.Sources[3] = baseYmlConfig;
					builder.Sources[4] = environmentYmlConfig;
					builder.Sources[5] = envConfig;
					builder.Sources[6] = cmdLineConfig;
```

File: docs/Architecture.dox [hunk 7]
```diff
@@ -23,7 +23,7 @@ This is a second process spawned by the Host Watchdog which facilitates the vast
...
+The @ref Tgstation.Server.Host.Core.Application class has two methods called by the framework. First the @ref Tgstation.Server.Host.Core.Application.ConfigureServices method sets up dependency injection of interfaces for Controllers, the @ref Tgstation.Server.Host.Database.DatabaseContext, and the component factories of the server. The framework handles constructing these things once the application starts. Configuration is loaded from the appropriate appSettings.yml into the @ref Tgstation.Server.Host.Configuration classes for injection as well. Then @ref Tgstation.Server.Host.Core.Application.Configure method is run which sets up the web request pipeline which currently has the following stack of handlers:
```
Comment (Cyberboss) [inline docs/Architecture.dox:26]: Also a minor nits

```suggestion
The @ref Tgstation.Server.Host.Core.Application class has two methods called by the framework. First the @ref Tgstation.Server.Host.Core.Application.ConfigureServices method sets up dependency injection of interfaces for Controllers, the @ref Tgstation.Server.Host.Database.DatabaseContext, and the component factories of the server. The framework handles constructing these things once the application starts. Configuration is loaded from the appropriate appsettings.yml into the @ref Tgstation.Server.Host.Configuration classes for injection as well. Then @ref Tgstation.Server.Host.Core.Application.Configure method is run which sets up the web request pipeline which currently has the following stack of handlers:
```

File: src/Tgstation.Server.Host/Tgstation.Server.Host.csproj [hunk 8]
```diff
@@ -114,11 +115,16 @@
     <None Include="ClientApp\package.json" />
     <None Include="ClientApp\tsconfig.json" />
     <ClientApp Include="ClientApp\src\**\*;ClientApp\public\**\*;ClientApp\tsconfig.json;ClientApp\package.json;ClientApp\package-lock.json" />
+    <None Remove="appsettings.yml" />
```
Comment (Cyberboss) [inline src/Tgstation.Server.Host/Tgstation.Server.Host.csproj:118]:
```suggestion
```

File: src/Tgstation.Server.Host/Tgstation.Server.Host.csproj [hunk 9]
```diff
@@ -114,11 +115,16 @@
...
     <AdditionalFiles Include="../../build/stylecop.json" />
   </ItemGroup>
 
   <ItemGroup>
-    <Content Update="appsettings.json">
+    <Content Include="appsettings.yml" />
+  </ItemGroup>
+
+  <ItemGroup>
```
Comment (Cyberboss) [inline src/Tgstation.Server.Host/Tgstation.Server.Host.csproj:126]:
```suggestion
```

**Review by Cyberboss — CHANGES_REQUESTED** [review]
Running this in the debugger shows that the setup wizards is actually triggering and crashing when trying to read from stdin. The issue is that the .ymls (base and production) are being added as the last configuration providers, giving them priority over env vars and the command line. Not good.

**Review by Cyberboss — CHANGES_REQUESTED** [review]
Remember to check a production build works after the basepath change (you can use the CI artifacts if you can build it locally). Also you still need to make `General:ConfigVersion` serialize as a semver string instead of an object.

**Review by Cyberboss — CHANGES_REQUESTED** [review]
Replace the environment .json exclusion in .gitignore with `/src/Tgstation.Server.Host/appsettings.*.*`

**Review by Cyberboss — CHANGES_REQUESTED** [review]
Also can you adjust the configuration provider order again. currently base-yaml overrides environment-json which is no beuno. The index of environment-json should be 2

**Review by Cyberboss — CHANGES_REQUESTED** [review]
Out of order still

File: src/Tgstation.Server.Host/Setup/SetupWizard.cs [hunk 10]
```diff
@@ -841,8 +841,11 @@ async Task<Uri> ParseAddress(string question)
...
+			var serializer = new SerializerBuilder().Build();
```
Comment (Cyberboss) [inline src/Tgstation.Server.Host/Setup/SetupWizard.cs:844]: Then you can add
```suggestion
			var serializer = new SerializerBuilder()
				.WithTypeConverter(new VersionConverter())
				.Build();
```

**Review by Cyberboss — CHANGES_REQUESTED** [review]
From https://www.cyotek.com/blog/using-custom-type-converters-with-csharp-and-yamldotnet-part-1

You need to implement a custom type converter for `System.Version` by creating a new class that implements `IYamlTypeConverter`.

Actually, since we already have the JSON equivalent, you can just slap that interface on [VersionConverter](https://github.com/tgstation/tgstation-server/blob/dev/src/Tgstation.Server.Host/Extensions/Converters/VersionConverter.cs)

**Review by Cyberboss — APPROVED** [review]

**Review by Cyberboss — APPROVED** [review]

**Comment by AffectedArc07** [comment]
It is my god given right to selfmerge

**Comment by tgstation-server** [comment]
Your Fix/Feature pull request delta is currently below zero (-9). Maintainers may close future Feature/Tweak/Balance PRs. Fixing issues or helping to improve the codebase will raise this score.

**Comment by AffectedArc07** [comment]
Oh go away

**Pull request merged**